module/VuFindSearch/src/VuFindSearch/Backend/Solr/QueryBuilder.php

```diff
@@ -180,9 +189,136 @@ public function build(AbstractQuery $query)
         }
         $params->set('q', $string);
 
+        // Handle any extra parameters:
+        foreach ($this->extraParams as $extraParam) {
+            if (empty($extraParam['param']) || empty($extraParam['value'])) {
+                continue;
+            }
+            if (!$this->checkParamConditions($query, $extraParam['conditions'] ?? [])
+            ) {
+                continue;
+            }
+            foreach ((array)$extraParam['value'] as $value) {
+                $params->add($extraParam['param'], $value);
+            }
+        }
+
         return $params;
     }
 
+    /**
+     * Check if the conditions match for an extra parameter
+     *
+     * @param AbstractQuery $query      Search query
+     * @param array         $conditions Required conditions
+     *
+     * @return bool
+     */
+    protected function checkParamConditions(
+        AbstractQuery $query,
+        array $conditions
+    ): bool {
+        if (empty($conditions)) {
+            return true;
+        }
+        $searchTypes = $this->getSearchTypes($query);
+        $searchTypeOk = null;
+        foreach ($conditions as $condition) {
+            if (is_array($condition)) {
+                $values = reset($condition);
+                $condition = key($condition);
+                switch ($condition) {
+                case 'SearchTypeIn':
+                    if (null === $searchTypeOk) {
```

Comment:
I'm a little confused about the interaction between SearchTypeIn and SearchTypeNotIn. I guess my head wants them to be either AND'ed together or OR'ed together, and I think what's actually happening is not quite either of those scenarios... is there a particular use case that led you to this implementation?

Reply:
It is a bit complicated because a single search may contain multiple search types, and I wanted to be able to do something like "Don't boost most recent material for PublishDate search unless there are AllFields or Title fields in the search as well." But I see it doesn't quite make sense. I've now simplified the rules and added AllSearchTypesIn as a condition that requires all search types to be in the given list. I think this would cover the use cases while being simple enough even for me to understand later on. :) But if you see more room for improvement, please let me know!